src/userdev/java/net/minecraftforge/gradle/userdev/util/Deobfuscator.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * ForgeGradle
+ * Copyright (C) 2018 Forge Development LLC
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+package net.minecraftforge.gradle.userdev.util;
+
+import net.minecraftforge.gradle.common.util.*;
+import org.apache.commons.io.IOUtils;
+import org.gradle.api.Project;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+public class Deobfuscator {
+    private final Project project;
+    private File cacheRoot;
+
+    public Deobfuscator(Project project, File cacheRoot) {
+        this.project = project;
+        this.cacheRoot = cacheRoot;
+    }
+
+    public File deobfBinary(File original, String mappings, String... cachePath) throws IOException {
+        project.getLogger().debug("Deobfuscating binary file {} with mappings {}", original.getName(), mappings);
+
+        File names = findMapping(mappings);
+        if (names == null || !names.exists()) {
+            return null;
+        }
+
+        File cacheDir = getOrCreateCacheDir(cachePath);
+
+        String mappedName = addMappingsToName(original.getName(), mappings);
```

Comment:
This is nothing but asking for collisions by not supporting unique identifiers, this is the main reason why I never bothered to add direct file dep support.
Everything should go through ivy/maven format.